flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java

```diff
@@ -0,0 +1,380 @@
+                a ->
+                        a.name(fieldName)
+                                .getter(getter)
+                                .setter((e, o) -> {})
+                                .attributeConverter(getAttributeConverter(typeInfo)));
+    }
+
+    private <AttributeT> AttributeConverter<AttributeT> getAttributeConverter(
+            TypeInformation<AttributeT> typeInfo) {
+        if (typeInfo instanceof BasicTypeInfo) {
+            return AttributeConverterProvider.defaultProvider()
+                    .converterFor(EnhancedType.of(typeInfo.getTypeClass()));
+        } else if (typeInfo.equals(PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO)) {
+            return getAttributeConverter(
+                    AttributeValueType.B,
+                    bytes ->
+                            bytes instanceof SdkBytes
+                                    ? AttributeValue.fromB((SdkBytes) bytes)
+                                    : AttributeValue.fromB(SdkBytes.fromByteArray((byte[]) bytes)));
+        } else if (typeInfo instanceof BasicArrayTypeInfo) {
+            BasicArrayTypeInfo<AttributeT, ?> basicArrayTypeInfo =
+                    (BasicArrayTypeInfo<AttributeT, ?>) typeInfo;
+            if (basicArrayTypeInfo.getComponentInfo().equals(BasicTypeInfo.STRING_TYPE_INFO)) {
+                return getAttributeConverter(
+                        AttributeValueType.SS,
+                        array -> AttributeValue.fromSs(Arrays.asList((String[]) array)));
+            } else if (basicArrayTypeInfo.getComponentInfo() instanceof NumericTypeInfo) {
+                return getAttributeConverter(
+                        AttributeValueType.NS,
+                        array ->
+                                AttributeValue.fromNs(
+                                        convertObjectArrayToStringList((Object[]) array)));
+            }
+
+            return new ArrayAttributeConverterProvider()
+                    .converterFor(EnhancedType.of(typeInfo.getTypeClass()));
+        } else if (typeInfo instanceof ObjectArrayTypeInfo) {
+            return getObjectArrayTypeConverter((ObjectArrayTypeInfo<AttributeT, ?>) typeInfo);
+        } else if (typeInfo instanceof PrimitiveArrayTypeInfo) {
+            PrimitiveArrayTypeInfo<AttributeT> primitiveArrayTypeInfo =
+                    (PrimitiveArrayTypeInfo<AttributeT>) typeInfo;
+            if (primitiveArrayTypeInfo.getComponentType() instanceof NumericTypeInfo) {
+                return getAttributeConverter(
+                        AttributeValueType.NS,
+                        array -> AttributeValue.fromNs(convertPrimitiveArrayToStringList(array)));
+            } else {
+                throw new IllegalArgumentException(
+                        String.format(
+                                "Unsupported primitive array typeInfo %s",
+                                primitiveArrayTypeInfo.getComponentType()));
+            }
```

Comment:
Have we considered using the `.attributeConverterProviders()` method in the table schema builder and reusing the constructs we use to DDB Table API?

```
TableSchema.builder(...)
                      .attributeConverterProviders(new ArrayAttributeConverterProvider(), ... (add any more here) ..., AttributeConverterProvider.defaultProvider());
```

We have an example here:
https://github.com/apache/flink-connector-aws/blob/main/flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/table/RowDataToAttributeValueConverter.java#L56-L72

Reply:
@hlteoh37 yes, the whole approach actually is very similar, however The way `AttributeConverterProvider` works is by mapping `EnhancedType` provided to an `AttributeConverter`, We here rely on Flink's `TypeInfo` rather than the `EnhancedType` that's why we created an alternative structure. Wdyt?

Reply:
Hm, ultimately `EnhancedType` is just a wrapper for the `AttributeConverterProvider` to match the right `AttributeConverter`. We cannot run away from this - in fact you have created a custom implementation of this logic in `getAttributeConverter`, which takes type information and constructs a custom attribute converter. 

Wouldn't using the `AttributeConverterProvider` more directly simplify the code?

Reply:
> We cannot run away from this 

This is not true, we are "running away from this" or rather replacing the mapping dependency on `EnhancedType` by directly attaching each resolved `AttribureConverter` to the attribute in
```
  typeInfo.getTypeClass(),
                a ->
                        a.name(fieldName)
                                .getter(getter)
                                .setter((e, o) -> {})
                                .attributeConverter(getAttributeConverter(typeInfo)))
```

> in fact you have created a custom implementation of this logic in getAttributeConverter, which takes type information and constructs a custom attribute converter.

Yes, I am aware of that and that exactly why using an attribute converter is against what we are trying to achieve here, Are you suggesting we don't depend on Flink's TypeInformation and resolve back to Java Classes?